pkg/storage/local.go

```diff
@@ -1510,6 +1510,84 @@ func (is *ImageStoreLocal) copyBlob(repo, blobPath, dstRecord string) (int64, er
 	return -1, zerr.ErrBlobNotFound
 }
 
+type blobStream struct {
+	reader io.Reader
+	closer io.Closer
+}
+
+func NewBlobStream(blobPath string, from, to int64) (io.ReadCloser, error) {
+	blobFile, err := os.Open(blobPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if from > 0 {
+		_, err = blobFile.Seek(from, io.SeekStart)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	blobstrm := blobStream{reader: blobFile, closer: blobFile}
+
+	blobstrm.reader = io.LimitReader(blobFile, (to-from)+1)
+
+	return &blobstrm, nil
+}
+
+func (bs *blobStream) Read(buf []byte) (int, error) {
+	return bs.reader.Read(buf)
+}
+
+func (bs *blobStream) Close() error {
+	return bs.closer.Close()
+}
+
+// GetBlobPartial returns a partial stream to read the blob.
+// blob selector instead of directly downloading the blob.
+func (is *ImageStoreLocal) GetBlobPartial(repo, digest, mediaType string, from, to int64,
+) (io.ReadCloser, int64, bool, error) {
+	var lockLatency time.Time
+
+	parsedDigest, err := godigest.Parse(digest)
+	if err != nil {
+		is.log.Error().Err(err).Str("digest", digest).Msg("failed to parse digest")
+
+		return nil, -1, true, zerr.ErrBadBlobDigest
+	}
+
+	blobPath := is.BlobPath(repo, parsedDigest)
+
+	is.RLock(&lockLatency)
+	defer is.RUnlock(&lockLatency)
+
+	binfo, err := os.Stat(blobPath)
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", blobPath).Msg("failed to stat blob")
+
+		return nil, -1, true, zerr.ErrBlobNotFound
+	}
+
+	if to < 0 || to > binfo.Size() {
+		to = binfo.Size()
+	}
+
+	blobReadCloser, err := NewBlobStream(blobPath, from, to)
+	if err != nil {
+		is.log.Error().Err(err).Str("blob", blobPath).Msg("failed to open blob")
+
+		return nil, -1, true, err
+	}
+
+	partial := true
+	if from == 0 || (to-from)+1 >= binfo.Size() {
```

Comment:
The `||` must be a `&&` otherwise range request like `0-100` are not partial which is wrong.
```suggestion
	if from == 0 && (to-from)+1 >= binfo.Size() {
```